drivers/flash/flash_hp_ra.c

```diff
@@ -0,0 +1,392 @@
+	.fsp_config = {
+		.data_flash_bgo = true,
+		.p_callback = bgo_callback,
+		.p_context = NULL,
+		.irq = (IRQn_Type)DT_INST_IRQ_BY_NAME(0, frdyi, irq),
+		.err_irq = (IRQn_Type)DT_INST_IRQ_BY_NAME(0, fiferr, irq),
+		.err_ipl = DT_INST_IRQ_BY_NAME(0, fiferr, priority),
+		.ipl = DT_INST_IRQ_BY_NAME(0, frdyi, priority),
+	}};
+
+#ifdef CONFIG_FLASH_EX_OP_ENABLED
+static int flash_ra_ex_op(const struct device *dev, uint16_t code, const uintptr_t in, void *out)
+{
+	int err = -ENOTSUP;
+
+	switch (code) {
+#if defined(CONFIG_FLASH_RA_WRITE_PROTECT)
+	case FLASH_RA_EX_OP_WRITE_PROTECT:
+		err = flash_ra_ex_op_write_protect(dev, in, out);
+		break;
+#endif /* CONFIG_FLASH_RA_WRITE_PROTECT */
+
+	default:
+		break;
+	}
+
+	return err;
+}
+#endif
+
+static int flash_ra_init(const struct device *dev)
+{
+	const struct device *dev_ctrl = DEVICE_DT_INST_GET(0);
+	const struct flash_hp_ra_config *cfg = dev->config;
+	struct flash_hp_ra_data *flash_data = dev->data;
+
+	if (!device_is_ready(dev_ctrl)) {
+		return -ENODEV;
+	}
+
+	if (cfg->flash_ra_parameters.write_block_size == UNIT_CODE_FLASH_PROGRAMMING) {
+		flash_data->FlashRegion = CODE_FLASH;
+	} else {
+		flash_data->FlashRegion = DATA_FLASH;
+	}
+
+	flash_data->controller = dev_ctrl->data;
+
+	return 0;
+}
+
+static void flash_controller_ra_irq_config_func(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	R_ICU->IELSR[DT_IRQ_BY_NAME(DT_DRV_INST(0), frdyi, irq)] = ELC_EVENT_FCU_FRDYI;
+	R_ICU->IELSR[DT_IRQ_BY_NAME(DT_DRV_INST(0), fiferr, irq)] = ELC_EVENT_FCU_FIFERR;
+
+	IRQ_CONNECT(DT_IRQ_BY_NAME(DT_DRV_INST(0), frdyi, irq),
+		    DT_IRQ_BY_NAME(DT_DRV_INST(0), frdyi, priority), fcu_frdyi_isr,
+		    DEVICE_DT_INST_GET(0), 0);
+	IRQ_CONNECT(DT_IRQ_BY_NAME(DT_DRV_INST(0), fiferr, irq),
+		    DT_IRQ_BY_NAME(DT_DRV_INST(0), fiferr, priority), fcu_fiferr_isr,
+		    DEVICE_DT_INST_GET(0), 0);
+
+	irq_enable(DT_INST_IRQ_BY_NAME(0, frdyi, irq));
+	irq_enable(DT_INST_IRQ_BY_NAME(0, fiferr, irq));
+}
+
+static int flash_controller_ra_init(const struct device *dev)
+{
+	fsp_err_t err = FSP_SUCCESS;
+	const struct flash_hp_ra_controller_config *cfg = dev->config;
+	struct flash_hp_ra_controller *data = dev->data;
+
+	cfg->irq_config(dev);
+
+	err = R_FLASH_HP_Open(&data->flash_ctrl, &data->fsp_config);
+	LOG_DBG("flash: open error=%d", err);
+
+	if (err != FSP_SUCCESS) {
+		return -EIO;
+	}
+
+	k_sem_init(&data->ctrl_sem, 1, 1);
+
+	return 0;
+}
+
+static struct flash_hp_ra_controller_config flash_hp_ra_controller_config = {
+	.irq_config = flash_controller_ra_irq_config_func,
+};
+
+#define RA_FLASH_INIT(index)                                                                       \
+	static struct flash_hp_ra_config flash_hp_ra_config_##index = {                            \
+		.flash_ra_parameters = {                                                           \
+			.write_block_size = DT_PROP(index, write_block_size),                      \
```

Comment:
You really want to take this parameter from the DTS?
As far as I can tell there is only valid value that the hardware specifies and if you set the wbs below that, writes will fails; there is no check whether the `write_block_size` is multiple of the hardware limitation for the parameter.
Note also that in `flash_ra_init` you use the `write_block_size` to decide whether a region is data or code; this means that if user is able to change these in DTS, without checks here, user may affect how region is considered just by incorrectly setting wbs.